### ISSUE TITLE
Source Control unsaved changes warning behaves unexpectedly

### ISSUE THREAD

**Issue by hrueger** [microsoft/vscode#190315]
Type: <b>Bug</b>

Hi,
I'm not sure if you can really call this a bug, but it certainly behaves different than one would think. Maybe it's worth an improvement.

Basically, I'm lazy and usually don't stage changes before I commit them using the source control panel in the sidebar. I've set the setting `Git: Enable Smart Commit` to `on`. It commits all changes if there are no staged changes.

When there is an unsaved file, I get a warning like this:
```
The following file has unsaved changes which won't be included in the commit if you proceed: [...]
Would you like to save it before committing?
```
![grafik](https://github.com/microsoft/vscode/assets/18656830/2c9feb52-219a-44df-9374-3836cd4a342f)

when I now press `Save All & Commit`, instead of adding this file to the commit, it only commits this one. This is probably due to that file being saved, staged and then commited. As there's now a staged file, the other ones are not committed.

Does that make sense?

Thanks in advance
Hannes

VS Code version: Code 1.81.1 (6c3e3dba23e8fadc360aed75ce363ba185c49794, 2023-08-09T22:22:42.175Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz (12 x 3696)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.96GB (4.79GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (60)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-sql-formatter|adp|1.4.4
ng-template|Ang|16.1.8
vscode-intelephense-client|bme|1.9.5
dart-code|Dar|3.70.0
flutter|Dar|3.70.0
mustache|daw|1.1.1
vscode-eslint|dba|2.4.2
FreeMarker|dco|0.0.9
devicescript-vscode|dev|2.14.14
gitlens|eam|14.2.1
prettier-vscode|esb|10.1.0
php-intellisense|fel|2.3.14
auto-close-tag|for|0.5.14
auto-rename-tag|for|0.1.10
copilot|Git|1.101.317
copilot-labs|Git|0.14.884
vscode-github-actions|git|0.26.1
vscode-pull-request-github|Git|0.70.0
hcl|has|0.3.2
Angular2|joh|16.0.1
vscode-json5|mrm|1.0.0
vscode-docker|ms-|1.26.0
vscode-language-pack-de|MS-|1.81.2023080209
vscode-dotnet-runtime|ms-|1.6.0
playwright|ms-|1.0.15
isort|ms-|2023.10.1
python|ms-|2023.14.0
vscode-pylance|ms-|2023.8.20
remote-containers|ms-|0.304.0
remote-ssh|ms-|0.102.0
remote-ssh-edit|ms-|0.86.0
remote-wsl|ms-|0.81.0
vscode-remote-extensionpack|ms-|0.24.0
cpptools|ms-|1.16.3
hexeditor|ms-|1.9.12
remote-explorer|ms-|0.4.1
remote-server|ms-|1.4.0
vsliveshare|ms-|1.0.5877
vscode-streamdeck|nic|4.1.6
pico-w-go|pau|3.2.1
vscode-versionlens|pfl|1.5.0
platformio-ide|pla|3.3.1
prisma|Pri|5.1.0
sqlite-viewer|qwt|0.3.13
vscode-xml|red|0.26.1
vscode-sort-json|ric|1.20.0
markdown-preview-enhanced|shd|0.6.8
svg-preview|Sim|2.8.3
vscode-stripe|str|2.0.14
svelte-vscode|sve|107.10.0
even-better-toml|tam|0.19.2
simple-php-formatter|tob|0.0.1
sort-lines|Tyr|1.10.2
vscode-counter|uct|3.2.1
adb-interface-vscode|vin|0.22.4
keyoti-changeallendoflinesequence|vs-|0.0.3
vscode-icons|vsc|12.5.0
html-css-class-completion|Zig|1.20.0
linkerscript|Zix|1.0.3
vscode-proto3|zxh|0.5.5

</details>

**Comment by Gobinath-B** [comment]
> Does that make sense?

@hrueger did you try the option above the Git:  Smart Commit Changes 

```
Git: Smart Commit Changes
Control which changes are automatically staged by Smart Commit.
```

This option is in  `all ` it will automatically stage all the changes.[default option]
if the option is in `tracked` it will automatically stage the tracked changes only.

![image](https://github.com/microsoft/vscode/assets/81898167/13a5adb3-0680-42a1-a2c9-59538bc31e7c)

**Comment by hrueger** [comment]
Hi @Gobinath-B
that setting is set to default (`all`).

**Comment by Gobinath-B** [comment]
Hello @hrueger 
Can you please try the alternative option that is `tracked` and let me know whether you still face this issue.

**Comment by hrueger** [comment]
Hi @Gobinath-B,
`tracked` shows the same problem (assuming none of the files I'm working with is untracked).

**Comment by Gobinath-B** [comment]
@hrueger , now i could reproduce this issue

> The following file has unsaved changes which won't be included in the commit if you proceed: [...]
Would you like to save it before committing?

Can you please explain this issue you mentioned step by step

> when I now press Save All & Commit, instead of adding this file to the commit, it only commits this one. This is probably due to that file being saved, staged and then commited. As there's now a staged file, the other ones are not committed.

**Comment by hrueger** [comment]
@Gobinath-B sure.
1. Enable `Git: Enable Smart Commit` in settings
2. Change a file and save it
3. Change another file, but don't save it
3. Input a commit message and press commit. Now, you're being asked if the unsaved file should be saved and included in the commit
4. If you press yes, you would expect it to be **included** in the commit together with the other (saved) files. However, it now only commits the previously unsaved file and not the other ones.

**Comment by Gobinath-B** [comment]
@hrueger thanks man.

Now i can reproduce the whole issue and totally understand the problem

**Comment by hrueger** [comment]
You're welcome 👍 Thanks for taking a look at that.

**Comment by Gobinath-B** [comment]
hello @lszomoru ,

I am super interesting to fix this bug, are you currently working on this issue?
if you are not, please notify me ASAP. I will work on this issue

**Comment by Gobinath-B** [comment]
@lszomoru I resolved this issue and raised a pull request.
Kindly review the PR.

**Comment by Gobinath-B** [comment]
@lszomoru I fixed this issue and raised a pull request  #190326 .Please review the code and merge it with main .

**Comment by lszomoru** [comment]
Closing this as duplicate of https://github.com/microsoft/vscode/issues/190657.
